horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -274,16 +275,52 @@ public Object getData(int id, String token, String schemaUri) {
    //this is nearly identical to TestServiceImpl.labelValues (except the return object)
    //this reads from the dataset table but provides data specific to the run...
    @Override
-   public List<ExportedLabelValues> labelValues(int runId, String filter, String sort, String direction, int limit, int page, List<String> include, List<String> exclude){
+   public List<ExportedLabelValues> labelValues(int runId, String filter, String sort, String direction, int limit, int page, List<String> include, List<String> exclude, boolean multiFilter){
       List<ExportedLabelValues> rtrn = new ArrayList<>();
       Run run = getRun(runId,null);
       if(run == null){
          throw ServiceException.serverError("Cannot find run "+runId);
       }
       Object filterObject = Util.getFilterObject(filter);
+      ObjectNode objectNode = null;
+
       String filterSql = "";
+      List<String> assumeMulti = new ArrayList<>();
+
       if(filterObject instanceof JsonNode && ((JsonNode)filterObject).getNodeType() == JsonNodeType.OBJECT){
-         filterSql = "WHERE "+TestServiceImpl.LABEL_VALUES_FILTER_CONTAINS_JSON;
+         objectNode = (ObjectNode) filterObject;
+         if(multiFilter) {
+            objectNode.fields().forEachRemaining(e -> {
+               String key = e.getKey();
+               JsonNode value = e.getValue();
+               if (value.getNodeType() == JsonNodeType.ARRAY) {//check if there are any matches
+                  ObjectNode arrayFilter = (ObjectNode) new JsonNodeFactory(false).objectNode().set(key, value);
+                  List<ExportedLabelValues> found = Collections.emptyList();
+                  found = labelValues(runId,arrayFilter.toString(),sort,direction,limit,page,include,exclude,false);
+                  if (found.isEmpty()) {
+                     assumeMulti.add(key);
+                  }
+               }
+            });
+            if (!assumeMulti.isEmpty()){
+               assumeMulti.forEach(objectNode::remove);
+            }
+         }
+         if(!objectNode.isEmpty()){
+            filterSql = "WHERE "+TestServiceImpl.LABEL_VALUES_FILTER_CONTAINS_JSON;
+         }
+         if(!assumeMulti.isEmpty()){
+            if(filterSql.isEmpty()){
+               filterSql = "WHERE ";
+            }
+            for(int i=0; i<assumeMulti.size(); i++){
+               if(!filterSql.endsWith("WHERE ")){
+                  filterSql+=" AND ";
+               }
+               filterSql += " jsonb_path_query_first(combined.values,CAST( :key"+i+" as jsonpath)) = ANY(:value"+i+") ";
+            }
+         }
+
```

Comment:
I see this new logic is duplicated in `TestServiceImpl` (the only difference I spot seems that, here, we are not defining `FILTER_PREFIX`), can we extract it in a common place to avoid duplication? I think this will make further fixes easier to do if we need to do in one single place ratehr than two.

Reply:
+1 I think we should avoid duplication

Reply:
I agree, the sql manipulation should be done in one place. The reason I didn't is because all of this code has to be replaced when we transition to composable labels. I originally only created the `multiFilter` for the `test` `labelValues` endpoint but was asked to also add it to `run` for "consistency". The expectation is that this code will be gone in a few weeks but that is probably overly optimistic. I'll refactor :)